### PR TITLE
Improved SSAO2 for sample count <16

### PR DESCRIPTION
**Pull request by fooware**
### Summary
This fix massively improves SSAO2 when using low sample counts, like the default 8.

### Background
It was discovered as part of PR #13621 but is submitted as a new PR since 13621 is now all about adding control over the SSAO2 denoiser without changing anything by default.

This small change however is so massively positive in all the tests I have done that it shouldn't be left out by default but be considered a bugfix.

### How it works
The fix is rather simple. When calculating if sample X is occluded by sample Y in the SSAO2 calculations we considered everything with a depth difference of 0 (zero) to mean Y occludes X. However, this does not take any calculation or precision errors into account, which causes samples along a "flat" surface to occlude each other. Basically, the flat surface was occluding itself...

This fix adds an epsilon of sorts to the calculation. The value 0.02 was selected from testing and gave the best improvements on all tests I did and at the same time practically left the 16-sample case untouched.

That said, on hardware that doesn't have the same precision in the shaders. might benefit from a higher threshold, as would some scenes like the kitchen example. But going over 0.02 starts to affect samples >=16 which isn't great from a backward compatibility standpoint.

I did not make this into a tunable setting, since I don't expect that a user of Babylon.js would understand how it affects the shader calculations. This should just work.

### Test results
In the screenshots below, the "8x" part denotes how many SSAO2 samples were used, old version is on the left and the new one is on the right.

Notice how the fixed version is pretty similar to the 16x reference when the sample count goes down. But the old version goes bananas with incorrect occlusions everywhere. Also note that the two versions at sample count 16 are basically identical.

The images were saved without the blur-stage of the SSAO2 filter active, to better show the SSAO2 calculations.

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by fooware** [comment]
@Popov72, could you please take a look at this one? It is fairly important IMHO.

**Comment by fooware** [comment]
Related issue: https://github.com/BabylonJS/Babylon.js/issues/12630

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/13652/merge

Test environment: 
https://babylonsnapshots.z22.web.core.windows.net/refs/pull/13652/merge/index.html

To test a playground add it to the URL, for example:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/13652/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/13652/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/13652/merge
https://gui.babylonjs.com/?snapshot=refs/pull/13652/merge
https://nme.babylonjs.com/?snapshot=refs/pull/13652/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/13652/merge#BCU1XR#0

File: packages/dev/core/src/Shaders/ssao2.fragment.fx [hunk 1]
```diff
@@ -98,13 +98,16 @@ float viewZToOrthographicDepth( in float viewZ, in float near, in float far ) {
...
 			// get sample linearDepth:
 		    float sampleDepth = abs(texture2D(depthSampler, offset.xy).r);
 			// range check & accumulate:
 		    difference = depthSign * samplePosition.z - sampleDepth;
 		    float rangeCheck = 1.0 - smoothstep(correctedRadius*0.5, correctedRadius, difference);
-		    occlusion += (difference >= 0.0 ? 1.0 : 0.0) * rangeCheck;
+
+			// We need to filter away diffs very close to 0 due to precision errors,
+			// otherwise we will get a ton of incorrect occlusions with low sample counts.
+		    occlusion += (difference >= 0.02 ? 1.0 : 0.0) * rangeCheck;
```
Comment (sebavan) [inline packages/dev/core/src/Shaders/ssao2.fragment.fx:110]: 0.02 looks like a magic number to me. Could this be configurable ?

**Comment by sebavan** [comment]
Just noticed @Popov72 comments and yup totally agree :-) making it configurable with default 0.02 would be great and lets merge it in the other PR ???

By the way, thanks a ton for all the improvments, it starts to look amazing !!!

**Comment by fooware** [comment]
The fix has now been moved to PR #13621 , closing this PR.

**Pull request closed without merging**